storage/src/main/java/com/tencent/rss/storage/handler/impl/HdfsClientReadHandler.java

```diff
@@ -63,96 +64,89 @@ public HdfsClientReadHandler(
-      return new ShuffleIndexResult(indexData);
-    } catch (Exception e) {
-      LOG.info("Fail to read index files for shuffleId[{}], partitionId[{}]", shuffleId, partitionId);
+      fs = ShuffleStorageUtils.getFileSystemForPath(baseFolder, hadoopConf);
+    } catch (IOException ioe) {
+      throw new RuntimeException("Can't get FileSystem for " + baseFolder);
     }
 
-    return new ShuffleIndexResult();
-  }
+    FileStatus[] indexFiles;
+    String failedGetIndexFileMsg = "Can't list index file in  " + baseFolder;
 
-  public ShuffleDataResult readShuffleData(ShuffleDataSegment shuffleDataSegment) {
-    int expectedLength = shuffleDataSegment.getLength();
-    if (expectedLength < 0) {
-      LOG.error("Invalid segment {}", shuffleDataSegment);
-      return null;
+    try {
+      // get all index files
+      indexFiles = fs.listStatus(baseFolder,
+          file -> file.getName().endsWith(Constants.SHUFFLE_INDEX_FILE_SUFFIX));
+    } catch (Exception e) {
+      LOG.error(failedGetIndexFileMsg, e);
+      return;
     }
 
-    byte[] data = currentDataReader.readData(shuffleDataSegment.getOffset(), expectedLength);
-    if (data.length != expectedLength) {
-      LOG.error("Fail to read expected[{}] data, actual[{}] and segment is {}",
-          expectedLength, data.length, shuffleDataSegment);
-      return null;
+    if (indexFiles != null && indexFiles.length != 0) {
+      for (FileStatus status : indexFiles) {
+        LOG.info("Find index file for shuffleId[" + shuffleId + "], partitionId["
+            + partitionId + "] " + status.getPath());
+        String filePrefix = getFileNamePrefix(status.getPath().toUri().toString());
+        try {
+          HdfsShuffleReadHandler handler = new HdfsShuffleReadHandler(filePrefix, readBufferSize, hadoopConf);
+          readHandlers.add(handler);
+        } catch (Exception e) {
+          LOG.warn("Can't create ShuffleReaderHandler for " + filePrefix, e);
+        }
+      }
+      readHandlers.sort(Comparator.comparing(HdfsShuffleReadHandler::getFilePrefix));
     }
-
-    return new ShuffleDataResult(data, shuffleDataSegment.getBufferSegments());
   }
 
+  // TODO: remove the useless segmentIndex
   @Override
   public ShuffleDataResult readShuffleData(int segmentIndex) {
-    if (segmentIndex >= shuffleDataSegments.size()) {
-      ShuffleIndexResult shuffleIndexResult = null;
-      while (readerIterator.hasNext()) {
-        Entry<String, HdfsFileReader> entry = readerIterator.next();
-        String key = entry.getKey();
-        currentIndexReader = entry.getValue();
-        if (currentIndexReader == null) {
-          LOG.error("Index reader of {} is invalid!", key);
-          continue;
-        }
-
-        currentDataReader = dataReaderMap.get(entry.getKey());
-        if (currentDataReader == null) {
-          LOG.error("Data reader of {} is invalid!", key);
-          continue;
-        }
+    if (readHandlerIndex >= readHandlers.size()) {
+      return null;
```

Comment:
return empty shuffleDataResult not null？

Reply:
OK